File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/practice/Keys and Rooms/bfs/Solution.java
import java.util.*;

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        bfs(rooms, visited, 0);
        for (boolean v : visited) {
            if (!v) {
                return false;
            }
        }

        return true;
    }

    private void bfs(List<List<Integer>> rooms, boolean[] visited, int start) {
        Deque<Integer> dq = new ArrayDeque<>();
        visited[start] = true;
        dq.offer(start);

        while (!dq.isEmpty()) {
            int current = dq.poll();
            for (int next : rooms.get(current)) {
                if (visited[next]) {
                    continue;
                }
                visited[next] = true;
                dq.offer(next);
            }
        }
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/practice/Keys and Rooms/bfs/solution.py
from collections import deque

class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False for _ in rooms]
        dq = deque()

        dq.append(0)
        visited[0] = True

        while dq:
            current = dq.popleft()
            for next in rooms[current]:
                if visited[next]:
                    continue
                
                dq.append(next)
                visited[next] = True

        return all(visited)
